source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
   return typeUrlPrefix() + "/" + qualified_name;
 }
 
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<int> ref_count_;
+  Buffer::InstancePtr buffer_;
+};
+
+static void derefBufferInstanceContainer(void* container_ptr) {
+  auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+  container->ref_count_--;
+  if (container->ref_count_ <= 0) {
+    delete container;
+  }
+}
+
+grpc::ByteBuffer Common::makeByteBuffer(Buffer::InstancePtr bufferInstance) {
+  if (!bufferInstance) {
+    return {};
+  }
+  Buffer::RawSlice oneRawSlice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int nSlices = bufferInstance->getRawSlices(&oneRawSlice, 1);
+  if (nSlices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{nSlices, std::move(bufferInstance)};
+  if (nSlices == 1) {
+    grpc::Slice oneSlice(oneRawSlice.mem_, oneRawSlice.len_, &derefBufferInstanceContainer,
+                         container);
+    return {&oneSlice, 1};
+  }
+  STACK_ARRAY(manyRawSlices, Buffer::RawSlice, nSlices);
+  container->buffer_->getRawSlices(manyRawSlices.begin(), nSlices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(nSlices);
+  for (int i = 0; i < nSlices; i++) {
+    slices.emplace_back(manyRawSlices[i].mem_, manyRawSlices[i].len_, &derefBufferInstanceContainer,
+                        container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments); }
+  std::atomic<int> ref_count_;
+  Buffer::BufferFragmentImpl* fragments = nullptr;
+  std::vector<grpc::Slice> slices_;
+};
+
+Buffer::InstancePtr Common::makeBufferInstance(const grpc::ByteBuffer& byteBuffer) {
+  auto buffer = std::make_unique<Buffer::OwnedImpl>();
+  if (byteBuffer.Length() == 0) {
+    return buffer;
+  }
+  // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
+  // lifetime of the Slice(s) exceeds our Buffer::Instance.
+  std::vector<grpc::Slice> slices;
+  byteBuffer.Dump(&slices);
+  if (slices.size() == 0) {
+    return buffer;
+  }
+  auto container = new ByteBufferContainer(static_cast<int>(slices.size()));
+  std::function<void(const void*, size_t, const Buffer::BufferFragmentImpl*)> releaser =
+      [container](const void*, size_t, const Buffer::BufferFragmentImpl*) {
+        container->ref_count_--;
+        if (container->ref_count_ <= 0) {
+          delete container;
+        }
+      };
+  // NB: addBufferFragment takes a pointer alias to the BufferFragmentImpl which is passed in so we
+  // need to ensure that the lifetime of those objects exceeds that of the Buffer::Instance.
+  container->fragments = static_cast<Buffer::BufferFragmentImpl*>(
```

Comment:
use `new []` (and `delete[]` in destructor), otherwise I think this may end up with alignment issue.

Reply:
::malloc guarantees that the result is appropriately aligned as per the standard.  delete[] is useful if one wishes to call the destructor of each object which is not the case here.

Reply:
I don't think malloc aligns C++ class well (which may have larger alignment than scalar types ), as it has no knowledge of it. We had similar issue in Lua that only happens in macOS release build.

Reply:
See https://stackoverflow.com/questions/8752546/how-does-malloc-understand-alignment

"I think the most relevant part of the Herb Sutter quote is the part I've marked in bold:

Alignment. Any memory Alignment. Any memory that's allocated dynamically via new or malloc is guaranteed to be properly aligned for objects of any type, but buffers that are not allocated dynamically have no such guarantee"

In particular all malloc() calls on 64-bit linux result in 16 byte aligned addresses.  If MacOS is not returning adequately aligned addresses from malloc() then it is a bug in the OS as the standard is specific in this regard.

That said, I am very curious about this Lua issue.  Could you forward me a pointer? If there is some issue with MacOS then of course we should work around it :)

Reply:
The problem is malloc doesn't understand C++ class alignment, which differs in platforms/toolchains (see example code below). In this case perhaps you can just avoid that by adding a couple of `static_assert`s with `alignof` and `sizeof` for `BufferFragmentImpl`, so it make sure each element is aligned in C++ requirement.

re: Lua issue is #6599 #5551 , the alignment issue can be reproduced by the following code: (build with `-fsanitize=address,undefined` so it always fail):

```C++
#include <iostream>

struct alignas(256) TestObject {
  int a;
};

int main() {
  TestObject *foo = static_cast<TestObject*>(malloc(sizeof(TestObject)));
  std::cout << foo << std::endl;
  std::cout << sizeof(TestObject) << std::endl;
  new (foo) TestObject;
}
```

Reply:
Ah, yes, if the class specifies a non-standard alignment (e.g. alignment for cache lines) then malloc will not understand it.  While that is not the case here I can see that it makes sense to future proof.

Reply:
Making this difficult is that I there is no default constructor (0 arg) for Buffer::BufferFragmentImpl so I went with alignof() and posix_memalign().  PTAL